quix-frontend/client/src/lib/sql-autocomplete/adapter/sql-autocomplete-adapter-utills.ts

```diff
@@ -0,0 +1,193 @@
+        }
+        if (typeof column === 'object' && typeof column.dataType === "object")  {
+          allChildren = allChildren.concat(getObjectChildren(column.dataType, column.name));
+        }
+  });
+}
+return allChildren;
+}
+
+export function getSearchCompletion(tables: TableInfo[] , prefix: string | undefined):any {
+  if(!prefix.includes('.')) {
+    return [];
+  }
+  const allChildren = getAllChildrenOfTables(tables);
+  const relevantPartOfPrefix = findRelevantPartOfPrefix(tables , prefix.split('.')).slice(0, -1);
+  // const relevantPartOfPrefix = findRelevantPartOfPrefix(tables , prefix.split('.')).slice(0, -1);
+  if( !relevantPartOfPrefix )  {
+    return []
+  }
+  const lastDotIndex = relevantPartOfPrefix.lastIndexOf('.');
+  const startOfSearch = lastDotIndex !== -1 ? relevantPartOfPrefix.slice(0, lastDotIndex + 1) : relevantPartOfPrefix;
+  const searchPart = relevantPartOfPrefix.replace(startOfSearch,'')
+  const filteredChildren = allChildren.filter(obj => {
+  const nameInLowerCase = obj.name.toLowerCase();
+  const parts = nameInLowerCase.split('.');
+    if (parts.length > 1) {
+        const substringAfterFirstDot = parts.slice(1).join('.');
+        const criteria = checkCriteria(substringAfterFirstDot , searchPart.toLowerCase());
+        const flterIfInDiffrenCollumn = obj.name.startsWith(relevantPartOfPrefix.split('.')[0]);
+        return obj.name.includes(startOfSearch) && substringAfterFirstDot.includes(searchPart.toLowerCase()) && criteria && flterIfInDiffrenCollumn ;
+    }
+    return false;
+});
+const prefixUntilLastDot = extractPrefixUntilLastDot(relevantPartOfPrefix) ;
+  const completionArray = filteredChildren.map(obj => ({
+    value: obj.name,
+    meta : typeof obj.dataType === 'object' ? 'row' : obj.dataType,
+    caption:  obj.name.slice( prefixUntilLastDot.length +1)
+  }));
+  return completionArray;
+}
+
+export function getNextLevel(tables: TableInfo[] , prefix: string | undefined): any {
+  const allChildren = getAllChildrenOfTables(tables);
+  // const relevantPartOfPrefix = findRelevantPartOfPrefix(tables , prefix.split('.')).slice(0, -1); //if same problem for both change in function itself
+  const relevantPartOfPrefix = findRelevantPartOfPrefix(tables , prefix.split('.')); //if same problem for both change in function itself
+  const relevantChildren = allChildren.filter(obj => {
+  const dotCount = obj.name.split('.').length - 1;
+    return obj.name.startsWith(relevantPartOfPrefix) && dotCount === relevantPartOfPrefix.split('.').length - 1;
+  });
+  const completionArray = relevantChildren.map(obj => ({
+    value: prefix.replace(relevantPartOfPrefix, '') + obj.name,
+    meta : typeof obj.dataType === 'object' ? 'row' : obj.dataType,
+    caption:  obj.name.slice(relevantPartOfPrefix.length)
+  }));
+  return completionArray
+}
+
+function checkCriteria(substringAfterFirstDot, searchPart) {
+  const lastDotIndex = substringAfterFirstDot.lastIndexOf('.');
```

Comment:
It seems like you have a lot of places that are doing the same non trivial stuff..
I see that you hace 4 places that looks like that:

`const lastDotIndex = someString.lastIndexOf('.');`

lets create a util function that do that and give redability everywhere 

`const lastDotIndex = getLastDotIndex(someString);`

serach for `.lastIndexOf('.')` in your project.. you will find at least 4 of them

Reply:
const lastDotIndex = findLastDotIndex(relevantPartOfPrefix);